Adopt reference implementation canonical json encoding

lukpueh opened the pull request:
Quick-fixes #11 

This fixes issues with compatibility with the Python reference implementation by porting the reference implementation json canonicalization function, based on the [OLPC specification](http://wiki.laptop.org/go/Canonical_JSON).

lukpueh commented:
Run below snippet in the root of the repo to quickly test cross-compatibility.

```shell
################################################
# REQUIES `maven` and `in-toto`
# MUST BE EXECUTED IN THE ROOT OF THIS REPO
################################################

# Init some variables
private_key="src/test/resources/somekey.pem"
public_key="src/test/resources/someotherkey.pem"
java_test="src/test/java/io/github/in_toto/TestCanonical.java"
link="teasdfst.link"

# Write test link file
cat >${link}<<EOL
{
 "signatures": [], "signed": {
    "_type": "link", "name": "test", "command": [],
    "byproducts": {},"products": {}, "materials": {},
    "environment": {
      "a": "WTF",
      "b": true,
      "c": false,
      "d": null,
      "e": 1,
      "f": "\u001b\uFFFFF\\\n\""
    }
  }
}
EOL

# Write java test file
cat >${java_test} <<EOL
import io.github.in_toto.models.Link;
import io.github.in_toto.keys.*;
import java.io.*;
import java.nio.file.*;
import org.junit.jupiter.api.Test;

class TestCanonical {
  @Test
  public void testCanonical () throws IOException {
    // Read, sign and save link
    Key privateKey = RSAKey.read("${private_key}");
    String linkString = new String(Files.readAllBytes(Paths.get("${link}")));
    Link link = Link.read(linkString);
    link.sign(privateKey);
    link.dump("${link}");
  }
}
EOL

# Run test (requires maven)
mvn -Dtest=TestCanonical test

# Verify link signature generated in java with Python (requires in-toto)
in-toto-sign -v --verify -f ${link} -k ${public_key}

# Clean up
rm ${link} ${java_test}
```

SantiagoTorres commented:
So I can confirm it works on my side after checking out your commit:

[![asciicast](https://asciinema.org/a/egjKGszWqpCIsAuuLcSNjMkyd.svg)](https://asciinema.org/a/egjKGszWqpCIsAuuLcSNjMkyd)

SantiagoTorres commented on src/main/java/io/github/in_toto/lib/JSONEncoder.java:
```diff
@@ -24,43 +25,129 @@
+            JsonObject obj = (JsonObject)src;
+            // Create an ordered list of the JsonObject's keys
+            TreeSet<String> keys = new TreeSet<>();
+            for (Map.Entry<String, JsonElement> entry : obj.entrySet()) {
```
Correct me if I'm wrong, but shouldn't these be ordered alphabetically?

lukpueh replied:
Good point. I actually reused the [code that was already there](https://github.com/in-toto/in-toto-java/pull/12/files/d84ecf62ad3ab087add0cdf3c3a92c93a9444935#diff-716698852f0323b14edbffd178b65cacL44), quickly checking the [`TreeSet` reference](https://docs.oracle.com/javase/7/docs/api/java/util/TreeSet.html#TreeSet()), which claims that elements are *"sorted according to the natural ordering of its elements"*, and thought that was okay.

Now double checking:
For `String` elements *natural ordering* means [*lexicographic ordering*](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#compareTo(java.lang.String)), which is the same what Python's `sorted` builtin does, which [we use in the reference implementation](https://github.com/secure-systems-lab/securesystemslib/blob/v0.11.2/securesystemslib/formats.py#L737). So we should be okay indeed, right?

SantiagoTorres replied:
Yes, thanks for double checking!

SantiagoTorres commented on src/main/java/io/github/in_toto/models/Link.java:
```diff
@@ -35,8 +35,8 @@
-            HashMap<String, String> environment, ArrayList<String> command,
-            HashMap<String, String> byproducts) {
+            HashMap<String, Object> environment, ArrayList<String> command,
+            HashMap<String, Object> byproducts) {
```
I'm unsure about these `String` to `Object`changes. Aren't we assuming that the `key:value` pairs are Strings? (you do claim that in the canonicalize function itself.

lukpueh replied:
I actually only claim that it's okay to assume that the [`HashMap` **keys** are strings](https://github.com/in-toto/in-toto-java/pull/12/files/d84ecf62ad3ab087add0cdf3c3a92c93a9444935#diff-716698852f0323b14edbffd178b65cacR86).

Here is what the specs say about [byproducts](https://github.com/in-toto/docs/blame/master/in-toto-spec.md#L927) and [environment](https://github.com/in-toto/docs/blame/master/in-toto-spec.md#L941):
>The `"byproducts"` field is an opaque dictionary
...
The format of the environment information is not mandated by the in-toto
specification

The metadata examples, on the other hand, show e.g. [`byproducts.return-value` as `string`](https://github.com/in-toto/docs/blame/v0.9/in-toto-spec.md#L1134), and [`environment.variables` as `list`](https://github.com/in-toto/docs/blame/master/in-toto-spec.md#L958-L965).

In order to provide cross-implementation compatibility (for signatures), we need to decide on a strict metadata format. We can either do this in the specification, or just internally for our different implementations. FWIW, in the reference implementation I currently let the json parser decide about the datatypes and that's also what I tried to do here.

What do you think?

SantiagoTorres replied:
I think this is ok. It's probably messy on the jenkins plugin side, but I imagine that'd require correcting on there...

SantiagoTorres commented on src/main/java/io/github/in_toto/models/Metablock.java:
```diff
@@ -80,7 +85,27 @@ public void dump(Writer writer)
+                // type of the contained values. Gson treats any numeric value
+                // where the type is not specified in the target data structure
+                // as double.
+                .registerTypeAdapter(Double.class, new JsonSerializer<Double>() {
```
I know this is "correct" Java, but i personally hate inline class declarations. can we give this a reasonable name? e.g., `ReasonableJsonDoubleSerializer` and declare it as a nested static(?) class or so?

SantiagoTorres replied:
Note, my e.g., is a **joke**

lukpueh replied:
Haha, sure, will do.

m0jt4b4 commented:
The verification fails in my system

[![asciicast](https://asciinema.org/a/4y0Cc73cyTdQQhZsoOq30D71V.svg)](https://asciinema.org/a/4y0Cc73cyTdQQhZsoOq30D71V)

lukpueh commented:
@SantiagoTorres, thanks for you review! I hope I addressed your concerns. The question about the data types of `environment` and `byproducts` remains open for discussion.

SantiagoTorres commented:
@lukpueh I wonder if we'd like to extend the test suite with some of these edgecases in the comments :+1:

lukpueh commented:
I added a test (in 94fca8e), which also made me discover a bug (fixed in e79f486). See commit messages for more details.

SantiagoTorres commented:
Thanks!
